### PR TITLE
[DMP 2026] feat: OCR Document Scanning & Demographic Auto-Fill (#718)

### LINKED ISSUE #718: [DMP 2026]: OCR Integration with Android Registration Client
## Description
This feature enables OCR (Optical Character Recognition) integration within the Android Registration Client to automatically extract demographic information from identity documents during registration.

Currently, operators manually enter demographic details such as name, address, and date of birth, which is time-consuming and error-prone. This feature introduces document scanning via the device camera, followed by OCR-based text extraction and automatic pre-filling of form fields.

The goal is to reduce manual effort, improve data accuracy, and accelerate the registration workflow, while maintaining high standards of data security and configurability.

## Goals
**Overall Goals**

1. Enable document capture using Android device camera
2. Integrate OCR engine (on-device/server-based) for text extraction
3. Map extracted text to demographic fields
4. Auto-fill demographic form fields with extracted data
5. Provide confidence scores and visual indicators
6. Support multilingual OCR extraction
7. Ensure secure handling of captured and extracted data
8. Allow configuration of OCR behavior without code changes

**Mid-Point Milestone**

1. Camera-based document capture with image quality guidance (alignment, blur, lighting)
2. Basic OCR integration (single language support)
3. Extraction of core fields (Name, DOB, Gender)
4. Auto-fill of extracted fields into demographic form
5. Manual edit capability for extracted values
6. Basic error handling (OCR failure fallback to manual entry)

### Setup/Installation

_No response_

### Expected Outcome

The final product will provide a seamless, assisted data entry experience for registration operators:

1. Operators can scan identity documents directly within the app
2. The system intelligently extracts relevant demographic data using OCR
3. Extracted values are automatically populated into the appropriate fields
4. Each field displays a confidence score with visual indicators (Green/Yellow/Red) to guide operator validation
5. Operators retain full control to review, edit, or override extracted data
6. The system supports multiple languages, adapting to deployment-specific requirements
7. Robust error handling ensures fallback to manual entry when needed
8. All data handling complies with security, encryption, and privacy standards
9. The feature is highly configurable (document types, OCR engine, thresholds, storage policies)

Overall, the system will significantly reduce registration time, minimize errors, and improve operator efficiency and applicant experience.

### Acceptance Criteria

AC1: Operator can capture a document photo using the Android Registration Client
AC2: OCR extracts text from the captured document and maps it to demographic fields
AC3: Extracted fields are pre-filled automatically and can be edited manually
AC4: System shows confidence scores and highlights low-confidence fields
AC5: OCR supports multiple languages based on deployment configuration
AC6: OCR errors are handled gracefully with recapture and manual entry fallback options
AC7: All extracted data complies with encryption and security standards of the registration client
AC8: OCR configurations (document types, thresholds, engine selection) are adjustable without code changes

### 5. Implementation Details

1. Architecture

- Modular OCR integration layer to support:
  - On-device OCR (e.g., ML Kit / Tesseract)
  - Server-based OCR (via API, configurable)
- Image capture module integrated with Android camera APIs
- Pre-processing pipeline for:
  - Image enhancement (denoising, contrast adjustment)
  - Orientation correction
2. Core Components
- Document Capture Module
  - Camera integration with real-time guidance (overlay UI for alignment)
  - Image quality validation (blur, glare detection)
- OCR Processing Engine
  - Pluggable OCR provider (on-device or server-based)
  - Language detection and multilingual extraction support
- Field Mapping Engine
  - Rule-based or template-based mapping of extracted text to:
    - Name
    - Address
    - DOB
    - Gender
    - Document Number
  - Configurable mapping rules
- Confidence Scoring
  - OCR engine-provided confidence scores
  - Threshold-based classification:
    - High (Green)
    - Medium (Yellow)
    - Low (Red)
- UI/UX Enhancements
  - Highlight low-confidence fields
  - Editable pre-filled fields
  - Error prompts and recapture options
- Error Handling
  - OCR failure detection
  - Partial extraction handling
  - Retry + fallback to manual entry

- Security & Privacy
  - End-to-end encryption of captured images and extracted data
  - Configurable storage policies for document images
  - No external data transfer unless server OCR is explicitly enabled
  - Compliance with MOSIP data protection standards
- Configurability
  - Feature flags for OCR enablement
  - Configurable:
    - OCR engine type (on-device/server)
    - Supported document types
    - Confidence thresholds
    - Language packs
    - Image storage policy
    
### Mockups/Wireframes

https://www.figma.com/design/88XguVxvtSyjjeAxQhlDvA/Android-Reg-Client?node-id=249-2&p=f&t=hFL8hqfErAsUalFf-0 

### Product Name

MOSIP - Android Registration Client

### Organisation Name

MOSIP

### Domain

Open Source Project

### Tech Skills Needed

Java

### Mentor(s)

Feature Support : @Pragya279 
Tech Support: @jainhitesh9998 

### Category

Other

### PR DESCRIPTION
**Pull request by Jay1q2w**
Resolves #718

### Summary
Introduces an end-to-end OCR (Optical Character Recognition) pipeline that enables operators to scan identity documents (PAN cards, passports, etc.) via the device camera or gallery upload, automatically extract demographic fields, and auto-fill them into the registration form, significantly reducing manual data entry and improving accuracy.

---

### Motivation
Manual data entry during registration is time-consuming and error-prone. This feature leverages on-device ML (Google ML Kit + TFLite) to:
- Reduce operator keystrokes by 60-80% for demographic fields
- Minimize transcription errors (name spellings, DOB, document numbers)
- Support both camera-based live scanning and gallery image upload
- Work fully offline via the local ML Kit provider (no server dependency)

---
### 1. High-Level End-to-End Pipeline

download

### 2. Camera Quality Check & Auto-Capture Lifecycle

download

### 3. Field Extraction & Normalization Flow
download

---

### Commit History (Feature Branch)

| # | Commit | Description |
|---|--------|-------------|
| 1 | `1b449a94` | Initial OCR module: CameraX capture, ML Kit integration, raw text extraction, field parsing |
| 2 | `a5f9852c` | CameraX frame preview, document upload pipeline, UI v1 |
| 3 | `e5034dd5` | UI v2 with enhanced DocAutoScan, demographic form auto-fill |
| 4 | `e4aecc8e` | Extraction logic refined — label priority, PAN card handling, header patterns |
| 5 | `f43c772` | Document Type Detection Resolved (Id_Classifier.tflite)+ confidence Scoring |

---

#### New Files — Native Android OCR Engine (26 files)
| File | Purpose |
|------|---------|
| `ocr/OcrPluginImpl.java` | Main plugin — orchestrates Flutter↔Native bridge, scan lifecycle, OCR pipeline |
| `ocr/OcrConfig.java` | Immutable config from `GlobalParamRepository` (enabled, provider, timeouts, quality thresholds) |
| `ocr/OcrParamKeys.java` | Central registry of all `mosip.registration.ocr.*` parameter keys |
| `ocr/DocumentScanner.java` | CameraX-based live scanner with auto-capture after 4s of sustained good quality |
| `ocr/QualityAnalyzer.java` | Per-frame brightness (Laplacian) + blur detection with motion detection |
| `ocr/ImagePreprocessor.java` | Grayscale → contrast boost → deskew (projection-profile) pipeline |
| `ocr/BitmapUtils.java` | CameraX `ImageProxy` → `Bitmap` conversion with rotation correction |
| `ocr/OcrAuditLogger.java` | Structured audit trail for OCR events (quality, success, errors) |
| `ocr/engine/OcrProvider.java` | Interface: `extract(bitmap, spec, callback) → Map<String, String>` |
| `ocr/engine/OcrEngine.java` | Interface: `extractText(bitmap) → raw text + confidence` |
| `ocr/engine/OcrProviderFactory.java` | Factory: selects MlKit or Remote, wraps in timeout decorator |
| `ocr/engine/MlKitOcrProvider.java` | Local OCR: classify → preprocess → ML Kit extract → field extraction |
| `ocr/engine/MlKitEngine.java` | Wraps Google ML Kit Text Recognition (Latin model, v16.0.0) |
| `ocr/engine/RemoteOcrProvider.java` | Server-side OCR: POST base64 image + field spec, parse JSON response |
| `ocr/engine/DocumentClassifier.java` | TFLite classifier: identifies document type (PAN, passport, etc.) |
| `ocr/engine/TimeoutEnforcingOcrProvider.java` | Decorator enforcing configurable timeout on any provider |
| `ocr/extraction/DemographicFieldExtractor.java` | **Core**: 3-pass extraction (geometry → fuzzy → regex) + value normalization |
| `ocr/extraction/ExtractionConfig.java` | Label maps, regex patterns, gender values, noise chars — server-overridable |
| `ocr/extraction/BlockGeometryParser.java` | Converts ML Kit `Text` (with bounding boxes) into layout-aware rows |
| `ocr/extraction/FuzzyMatcher.java` | Jaro-Winkler similarity for OCR-noisy label matching (threshold: 0.82) |
| `ocr/extraction/DateNormalizer.java` | Normalizes OCR dates to `yyyy/MM/dd` (9 formats + OCR char substitution) |
| `ocr/extraction/TextNormalizer.java` | Unicode NFKC, noise removal, multi-space collapse, blood group normalization |
| `ocr/extraction/OcrUiSpecProvider.java` | Reads field spec from `GlobalParamRepository` → `List<FieldSpec>` |
| `ocr/models/FieldSpec.java` | DTO: field id, type, controlType, subType |
| `ocr/models/OcrErrorCode.java` | Enum: 11 error codes (101-111) with retryability classification |
| `ocr/models/OcrError.java` | Client-side error model for camera/quality failures |

#### New Files — Flutter/Dart UI (4 files)
| File | Purpose |
|------|---------|
| `lib/provider/ocr_scan_provider.dart` | State machine (idle→scanning→processing→success/error), field mapping, transliteration |
| `lib/ui/scanner/ocr_scan_page.dart` | Full-screen camera scanner UI with reticle, quality guidance, auto-fill on success |
| `lib/ui/scanner/ocr_upload_page.dart` | Gallery upload alternative with preview + scan trigger |
| `lib/model/ocr_error_messages.dart` | Human-readable error message lookup for all OCR error codes |

#### New Files — Assets & Config (2 files)
| File | Purpose |
|------|---------|
| `android/app/src/assets/Id_Classifier.tflite` | TFLite document classification model (13MB) |
| `android/app/src/assets/labels.txt` | Classifier labels |

#### Modified Files — Build & Config
| File | Change |
|------|--------|
| `android/app/build.gradle` | Added ML Kit, TFLite, CameraX, Guava deps; `noCompress "tflite"`; removed SonarQube |
| `android/app/src/main/AndroidManifest.xml` | Added CAMERA permission, camera features, ML Kit OCR dependency declaration |
| `pubspec.yaml` | Bumped version 1.0.1→1.1.1, SDK constraint widened to `<4.0.0`, added `camera` dep |

#### Modified Files — Flutter UI Integration
| File | Change |
|------|--------|
| `lib/ui/process_ui/widgets/generic_process_screen_content.dart` | Added OCR auto-fill bar (Scan + Upload buttons) on Demographics screen |
| `lib/ui/process_ui/widgets/gender_control.dart` | OCR-aware re-sync: preserves OCR-written values, re-reads from fieldInputValue |
| `lib/ui/process_ui/widgets/age_date_control.dart` | OCR-aware re-sync: parses OCR dates into date/age controllers |
| `lib/ui/process_ui/widgets/textbox_control.dart` | Reads initial values from fieldInputValue via controllerMap |
| `lib/ui/process_ui/widgets/radio_button_control.dart` | Listens for OCR writes, updates selection by code/name match |
| `pigeon/ocr_messages.dart` | Pigeon interface: `OcrHostApi` + `OcrFlutterApi` contract |

#### Modified Files — Services
| File | Change |
|------|--------|
| `AuditEvent.java` | Added 5 OCR audit events: REG-EVT-106, 115, 116, 117, 118 |
#### OCR Audit Event Registry
| Event ID | Event Enum | Event Type | Description | Stored Description / Payload Arguments |
|:---:|:---:|:---:|---|---|
| `REG-EVT-106` | `DOCUMENT_SCAN` | `USER_EVENT` | Operator tapped the "Scan" button on the Demographics form | Emits default event description: *"Scan document button clicked"* |
| `REG-EVT-115` | `OCR_SCAN_INITIATED` | `USER_EVENT` | Camera OCR scanning session launched | Emits default event description: *"OCR scan page opened for document capture"* |
| `REG-EVT-116` | `OCR_UPLOAD_INITIATED` | `USER_EVENT` | Document upload page opened for gallery picking | Emits default event description: *"OCR upload page opened for document image selection"* |
| `REG-EVT-117` | `OCR_SCAN_SUCCESS` | `USER_EVENT` | OCR extraction completed successfully | Parameterized payload: <br>• `param[0]`: `documentType` (e.g., `"pan_card_front"` or `"unknown"`)<br>• `param[1]`: `fieldsExtractedCount` (e.g., `"4"`)<br>• `param[2]`: `confidence` (e.g., `"88.0"` or `"N/A"`) |
| `REG-EVT-118` | `OCR_FIELDS_APPLIED` | `USER_EVENT` | Extracted OCR fields auto-filled into form controls | Parameterized payload: <br>• `param[0]`: `documentType` (e.g., `"pan_card_front"` or `"unknown"`)<br>• `param[1]`: `appliedCount` (e.g., `"4"`) |

---

### Key Design Decisions

1. **Dual OCR Provider Strategy** — Local (ML Kit) for offline capability + Remote (HTTP) for server-side processing. Selected via `mosip.registration.ocr.provider` global param.

2. **3-Pass Extraction Pipeline** — Geometry-aware (bounding box layout) → Fuzzy matching (Jaro-Winkler) → Regex fallback. This handles OCR noise, tilted scans, and varying document layouts.

3. **Timeout Enforcement** — `TimeoutEnforcingOcrProvider` decorator ensures the UI never hangs, regardless of provider implementation.

4. **Generation Counter** — `AtomicInteger scanGeneration` in `OcrPluginImpl` prevents stale async results when scans are rapidly restarted.

5. **Server-Overridable Configuration** — All extraction labels, quality thresholds, and field regexes are stored in `GlobalParamRepository` and can be updated server-side without app releases.

---

#### Prerequisites
- Android device with camera (emulator camera not recommended for quality analysis testing)
- App built with: `flutter build apk --debug`
- Server running with OCR global params configured (or defaults will be used) (OR OCRTestDataSeeder.java in android\app\src\main\java\io\mosip\registration_client\ocr\OcrTestDataSeeder.java)

#### Test Steps

**Camera Scan Flow:**
1. Launch the app → Login → Start a new registration → Navigate to Demographics
2. Observe the OCR auto-fill bar with "Scan" and "Upload" buttons (visible when `mosip.registration.ocr.enabled = true`)
3. Tap **"Scan"** → Grant camera permission if prompted
4. Point camera at an identity document
5. Observe real-time quality guidance (brightness/blur feedback)
6. Auto-capture triggers after ~4 seconds of good quality (or tap shutter to force)
7. Observe processing overlay → Success: fields auto-filled into form → Snackbar shows count
8. Verify extracted fields: name, DOB, gender, document number, etc.

**Gallery Upload Flow:**
1. Tap **"Upload"** on the Demographics screen
2. Select an image from gallery
3. Preview shown → Tap **"Scan Document"**
4. Same extraction pipeline runs → Fields auto-filled

**Edge Cases:**
- Test with blurry/dark images → Should show quality guidance, not capture
- Test with rotated document → ImagePreprocessor deskews up to ±10°

---

### Dependencies Added

| Library | Version | Purpose |
|---------|---------|---------|
| `com.google.mlkit:text-recognition` | 16.0.0 | On-device OCR text extraction |
| `org.tensorflow:tensorflow-lite` | 2.14.0 | Document classification model |
| `org.tensorflow:tensorflow-lite-support` | 0.4.4 | TFLite image preprocessing |
| `androidx.camera:camera-*` | 1.3.0 | CameraX preview + image analysis |
| `com.google.guava:guava` | 31.1-android | ListenableFuture for CameraX |

---

#### If not in Local DB then Default values used

| Parameter Key | Default | Type | Description |
|---|:---:|:---:|---|
| `mosip.registration.ocr.enabled` | `false` | `Boolean` | Master toggle for the OCR scan & upload feature |
| `mosip.registration.ocr.provider` | `"mlkit"` | `String` | Provider selection: `"mlkit"` (on-device) or `"remote"` (HTTP REST API) |
| `mosip.registration.ocr.response.timeout` | `10000` | `Integer` | Max OCR extraction timeout in milliseconds |
| `mosip.registration.ocr.service.url` | `""` | `String` | Endpoint URL for external OCR service (when provider is `"remote"`) |
| `mosip.registration.ocr.quality.max_retries` | `8` | `Integer` | Max consecutive bad frames before unlocking the "Capture Anyway" override |
| `mosip.registration.ocr.quality.thresholds.brightness_min` | `80.0` | `Float` | Min average frame luminance (below this triggers *"Too dark"*) |
| `mosip.registration.ocr.quality.thresholds.brightness_max` | `180.0` | `Float` | Max average frame luminance (above this triggers *"Too bright"*) |
| `mosip.registration.ocr.quality.thresholds.blur_variance` | `300.0` | `Double` | Min Laplacian variance threshold (below this triggers *"Hold steady"*) |
| `mosip.registration.ocr.ui.spec` | *(JSON)* | `String` | Demographic field specifications with `ocrKey` and schema mappings |
| `mosip.registration.ocr.extraction.labels` | *(JSON)* | `String` | Configurable label dictionaries mapped per field/subType |
| `mosip.registration.ocr.extraction.gender_values` | *(JSON)* | `String` | Whitelist of recognized gender tokens for normalization |
| `mosip.registration.ocr.extraction.month_names` | *(JSON)* | `String` | Month name mappings across supported languages for date normalization |
| `mosip.registration.ocr.extraction.header_patterns` | *(JSON)* | `String` | Stopping boundaries/headers (e.g., Government of India, Tax Dept) |
| `mosip.registration.ocr.extraction.noise_chars` | `":;—–/\\|"` | `String` | Punctuation/delimiter characters stripped during token extraction |
| `mosip.registration.ocr.extraction.fuzzy_threshold` | `0.82` | `Double` | Jaro-Winkler similarity threshold for Pass 2 fuzzy label matching |
| `mosip.registration.ocr.extraction.field_regexes` | *(JSON)* | `String` | Regex pattern dictionary for Pass 3 fallback (DOB, email, phone, PIN) |

> [Note]
> ### Current Work-in-Progress (WIP) Note: Improving Extraction Efficiency
> * **Status:** Currently I am improving the regex matching and fuzzy efficiency of the key value extraction

## Summary by CodeRabbit

* **New Features**
  * Added camera-based OCR scanning with document framing, quality guidance, automatic capture, and manual capture.
  * Added OCR image upload with preview, processing status, retry, and manual-entry options.
  * Extracted document fields can populate registration forms, including names, dates, gender, and other demographic data.
  * Added local and remote OCR processing, document recognition, user-friendly error messages, and audit events.

* **Bug Fixes**
  * Improved transliteration fallback behavior and synchronization of externally updated form values.
  * Improved OCR handling for image quality, date normalization, field matching, and processing timeouts.

### CODE REVIEW

**Comment by Jay1q2w** [comment]
Document-type classification has been completed and integrated into the OCR pipeline using Id_Classifier.tflite, with confidence scores returned alongside the predicted document type.

The classifier currently supports:

>aadhar_back
>aadhar_front
>driving_license_back
>driving_license_front
>pan_card_front
>passport
>voter_id

File: macos/Flutter/GeneratedPluginRegistrant.swift [hunk 1]
```diff
@@ -8,6 +8,7 @@ import Foundation
 import connectivity_plus
 import file_selector_macos
 import flutter_secure_storage_macos
+import geolocator_apple
```
Comment (GOKULRAJ136) [inline macos/Flutter/GeneratedPluginRegistrant.swift:11]: This file is auto-generated by flutter pub get and has nothing to do with OCR (the change is geolocator registration). Please revert it.

File: windows/flutter/generated_plugin_registrant.cc [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #include <connectivity_plus/connectivity_plus_windows_plugin.h>
 #include <file_selector_windows/file_selector_windows.h>
 #include <flutter_secure_storage_windows/flutter_secure_storage_windows_plugin.h>
+#include <geolocator_windows/geolocator_windows.h>
```
Comment (GOKULRAJ136) [inline windows/flutter/generated_plugin_registrant.cc:12]: Auto-generated from your local build and not related to OCR. Please revert it

File: windows/flutter/generated_plugins.cmake [hunk 3]
```diff
@@ -6,6 +6,7 @@ list(APPEND FLUTTER_PLUGIN_LIST
   connectivity_plus
   file_selector_windows
   flutter_secure_storage_windows
+  geolocator_windows
```
Comment (GOKULRAJ136) [inline windows/flutter/generated_plugins.cmake:9]: Auto-generated from your local build and not related to OCR. Please revert it

File: pubspec.lock [hunk 4]
```diff
@@ -129,6 +129,46 @@ packages:
       url: "https://pub.dev"
     source: hosted
     version: "8.11.1"
+  camera:
```
Comment (GOKULRAJ136) [inline pubspec.lock:132]: These lock changes come from your local pub get. The only OCR-related entry is camera, which should be removed along with the unused dependency. Please revert this aswell.

File: test/login_test.mocks.dart [hunk 5]
```diff
@@ -0,0 +1,162 @@
+// Mocks generated by Mockito 5.4.2 from annotations
```
Comment (GOKULRAJ136) [inline test/login_test.mocks.dart:1]: This is a Mockito file generated by build_runner, and master doesn't track mocks. It's also unrelated to OCR. Please remove it from the PR.

File: .fvmrc [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (GOKULRAJ136) [inline .fvmrc:1]: This is your local FVM setup, and the Flutter version is already pinned in CI and the README. Please remove it from the PR.

File: .gitignore [hunk 7]
```diff
@@ -88,3 +88,22 @@ app.*.map.json
 /ios/Runner/pigeon.h
 /ios/Runner/pigeon.m
 
+
```
Comment (GOKULRAJ136) [inline .gitignore:91]: Please revert the .gitignore changes:

- The .fvm/ entry is local tooling. Keep it in your global gitignore or .git/info/exclude.

- The Pigeon bridge entries (L97, L100) stop the build from working on a clean checkout, unless pigeon.sh generates those files.

- *.tflite+ has a typo and contradicts the committed model.

- Ignoring OcrTestDataSeeder.java means reviewers can't follow the test steps.

Only add an entry here if the whole project needs it.

File: android/app/src/main/java/io/mosip/registration_client/ocr/engine/MlKitOcrProvider.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        Bitmap processed;
+        try {
+            processed = ImagePreprocessor.prepare(bitmap);
+        } catch (Exception e) {
+            Log.w(TAG, "ImagePreprocessor failed, falling back to raw bitmap", e);
+            processed = bitmap;
+        }
+
+        final String finalDocumentType   = documentType;
+        final float  finalConfidence     = classificationConfidence;
```
Comment (GOKULRAJ136) [inline android/app/src/main/java/io/mosip/registration_client/ocr/engine/MlKitOcrProvider.java:80]: The confidence sent to Flutter and shown to the operator is the TFLite document-type score, not the OCR confidence. The ML Kit `avgConfidence` (L87) is only logged. AC4 needs OCR confidence, and it needs to be per field, so the operator knows which values to re-check.

File: android/app/src/main/java/io/mosip/registration_client/ocr/engine/MlKitEngine.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+public class MlKitEngine implements OcrEngine {
+
+    public interface RichTextCallback {
+        void onSuccess(@NonNull Text visionText, float avgConfidence);
+        void onFailure(@NonNull String errorCode, @NonNull String message, boolean isRetryable);
+    }
+    private TextRecognizer recognizer;
+
+    public MlKitEngine() {
+        this.recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);
```
Comment (GOKULRAJ136) [inline android/app/src/main/java/io/mosip/registration_client/ocr/engine/MlKitEngine.java:28]: This is the Latin-only recognizer, so AC5 (multilingual per deployment config) isn't met. Please make the recognizer or script configurable, and document which languages are supported.

File: android/app/src/main/java/io/mosip/registration_client/ocr/OcrPluginImpl.java [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+            @NonNull FlutterEngine flutterEngine,
+            @NonNull Activity activity,
+            @NonNull GlobalParamRepository globalParamRepository) {
+
+        OcrFlutterApi flutterApi = new OcrFlutterApi(
+                flutterEngine.getDartExecutor().getBinaryMessenger());
+
+        // OcrTestDataSeeder.seedIfEmpty(globalParamRepository);  //ONLY FOR TESTING
+
+        OcrConfig config = OcrConfig.from(globalParamRepository);
```
Comment (GOKULRAJ136) [inline android/app/src/main/java/io/mosip/registration_client/ocr/OcrPluginImpl.java:91]: Config and provider are built once, at app startup. On a fresh install, global params aren't synced yet, so OCR stays disabled until the app restarts. Any server-side change also needs a restart. Please read the config at scan start, or reload it after sync. This affects AC8.

File: android/app/src/main/java/io/mosip/registration_client/ocr/extraction/ExtractionConfig.java [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+                "jul", "aug", "sep", "sept", "oct", "nov", "dec",
+                "january", "february", "march", "april", "june", "july",
+                "august", "september", "october", "november", "december"
+        ));
+        return loadSetFromParam(repo, OcrParamKeys.OCR_EXTRACTION_MONTH_NAMES, defaults);
+    }
+
+    private Set<String> loadHeaderPatterns(GlobalParamRepository repo) {
+        Set<String> defaults = new HashSet<>(Arrays.asList(
+                "government of", "republic of", "income tax", "income tax department",
```
Comment (GOKULRAJ136) [inline android/app/src/main/java/io/mosip/registration_client/ocr/extraction/ExtractionConfig.java:186]: Country-specific header patterns and regexes shouldn't be hardcoded defaults. Please keep the defaults generic and let deployments supply these rules through global params.

File: lib/ui/scanner/ocr_upload_page.dart [hunk 12]
```diff
@@ -0,0 +1,832 @@
...
+  // ---------------------------------------------------------------------------
+  // File Picker
+  // ---------------------------------------------------------------------------
+  Future<void> _openPicker() async {
+    if (_pickerActive) return;
+    setState(() => _pickerActive = true);
+
+    try {
+      final ImagePicker picker = ImagePicker();
+      final XFile? picked = await picker.pickImage(
```
Comment (GOKULRAJ136) [inline lib/ui/scanner/ocr_upload_page.dart:129]: image_picker writes a re-encoded copy of the ID document into the app cache, and it's never deleted. Please delete it after processing, on success, error and cancel. This relates to AC7.
